pkg/operation/shoot/shoot.go

```diff
@@ -262,6 +264,35 @@ func (s *Shoot) GetInfo() *gardencorev1beta1.Shoot {
 	return s.info.Load().(*gardencorev1beta1.Shoot)
 }
 
+// GenerateMachineDeployments creates a slice of extensionsv1alpha1.MachineDeployment using the worker spec present
+// in the provider section of shoot.Spec. This method is only used for generating the slice of machine deployments
+// required by cluster autoscaler to populate its flags.
+func (s *Shoot) GenerateMachineDeployments() []extensionsv1alpha1.MachineDeployment {
+	shoot := s.GetInfo()
+	var machineDeployents []extensionsv1alpha1.MachineDeployment
+	for _, pool := range shoot.Spec.Provider.Workers {
+		zoneLen := int32(len(pool.Zones))
+		// for a worker with no zone there will only be one machine deployment and it will not have a `-z1` suffix in its name.
+		if zoneLen == 0 {
+			machineDeployents = append(machineDeployents, extensionsv1alpha1.MachineDeployment{
+				Name:    fmt.Sprintf("%s-%s", s.SeedNamespace, pool.Name),
+				Minimum: pool.Minimum,
+				Maximum: pool.Maximum,
+			})
+		} else {
+			for zoneIndex := range pool.Zones {
+				zoneIdx := int32(zoneIndex)
+				machineDeployents = append(machineDeployents, extensionsv1alpha1.MachineDeployment{
+					Name:    fmt.Sprintf("%s-%s-z%d", s.SeedNamespace, pool.Name, zoneIndex+1),
+					Maximum: worker.DistributeOverZones(zoneIdx, pool.Maximum, zoneLen),
+					Minimum: worker.DistributeOverZones(zoneIdx, pool.Minimum, zoneLen),
+				})
```

Comment:
Unfortunately, we cannot assume that the final `MachineDepoyment`s will look the result above since it's the decision of the provider extension.

Example:
- Provider AWS name calculation
https://github.com/gardener/gardener-extension-provider-aws/blob/b94917b0149e8bb2b2bfe16881a60a4a0a12f03f/pkg/controller/worker/machines.go#L189

```
deploymentName          = fmt.Sprintf("%s-%s-z%d", w.worker.Namespace, pool.Name, zoneIndex+1)
```

- Provider Metal-Stack name calculation:
https://github.com/metal-stack/gardener-extension-provider-metal/blob/e48d3e9590bf575f5617284c1ee171fc2bf6031b/pkg/controller/worker/machines.go#L183


```
deploymentName          = fmt.Sprintf("%s-%s", w.worker.Namespace, pool.Name)
```